zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -0,0 +1,532 @@
+            }
+        }
+
+        assert_eq!(
+            word_ranges.len(),
+            self.num_word,
+            "not enough rectangle regions for words"
+        );
+
+        let (cell_idxs, unused_idxs) = cell_idxs.split_at(self.num_cell);
+        (word_ranges, cell_idxs.to_vec(), unused_idxs.to_vec())
+    }
+}
+
+#[derive(Debug)]
+struct CaseAllocation<F> {
+    selector: Cell<F>,
+    words: Vec<Word<F>>,
+    cells: Vec<Cell<F>>,
+    resumption: Option<Resumption<F>>,
+}
+
+#[derive(Clone, Debug)]
+struct Resumption<F> {
+    caller_id: Cell<F>,
+    gas_available: Cell<F>,
+}
+
+impl<F: FieldExt> Resumption<F> {
+    fn new(cells: &[Cell<F>]) -> Self {
+        assert_eq!(cells.len(), NUM_CELL_RESUMPTION);
+
+        Self {
+            caller_id: cells[0].clone(),
+            gas_available: cells[1].clone(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct OpExecutionState<F> {
+    pub is_executing: Cell<F>,
+    pub global_counter: Cell<F>,
+    pub call_id: Cell<F>,
+    pub program_counter: Cell<F>,
+    pub stack_pointer: Cell<F>,
+    gas_counter: Cell<F>,
+    opcode: Cell<F>,
+}
+
+impl<F: FieldExt> OpExecutionState<F> {
+    pub(crate) fn new(cells: &[Cell<F>]) -> Self {
+        assert_eq!(cells.len(), NUM_CELL_OP_EXECTION_STATE);
+
+        Self {
+            is_executing: cells[0].clone(),
+            global_counter: cells[1].clone(),
+            call_id: cells[2].clone(),
+            program_counter: cells[3].clone(),
+            stack_pointer: cells[4].clone(),
+            gas_counter: cells[5].clone(),
+            opcode: cells[6].clone(),
+        }
+    }
+}
+
+trait OpGadget<F: FieldExt> {
+    const RESPONSIBLE_OPCODES: &'static [u8];
+
+    const CASE_CONFIGS: &'static [CaseConfig];
+
+    fn construct(case_allocations: Vec<CaseAllocation<F>>) -> Self;
+
+    fn constraints(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+    ) -> Vec<Constraint<F>>;
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error>;
+}
+
+// Preset stores default values for each case of op gadget
+#[derive(Clone, Default)]
+struct Preset<F> {
+    qs_byte_lookups: [F; CIRCUIT_HEIGHT],
+    free_cells: Vec<(usize, F)>,
+}
+
+#[derive(Clone)]
+pub(crate) struct OpExecutionGadget<F> {
+    r: F,
+    qs_byte_lookups: Vec<Cell<F>>,
+    state_curr: OpExecutionState<F>,
+    state_next: OpExecutionState<F>,
+    qs_ops: Vec<Cell<F>>,
+    free_cells: Vec<Cell<F>>,
+    resumption: Resumption<F>,
+    // TODO: use OpcodeId from bus_mapping
+    qs_op_idx_map: HashMap<u8, usize>,
+    preset_map: HashMap<(usize, Case), Preset<F>>,
+    add_gadget: AddGadget<F>,
+    push_gadget: PushGadget<F>,
+}
+
+impl<F: FieldExt> OpExecutionGadget<F> {
+    // TODO: refactor input type
```

Comment:
Can we file an issue for that?

Reply:
Sure, there are also many other places using many inputs like this, we should probably group some of them in a more readable way.

Reply:
Sounds nice. Let's do so before merging!

Reply:
Mentioned in #74